tests/lca/imagebasedupgrade/cnf/upgrade-talm/tests/rollback-after-failed-upgrade-test.go

```diff
@@ -0,0 +1,129 @@
+					tsparams.IbuCguNamespace)
+				Expect(err).ToNot(HaveOccurred(), "Failed to delete upgrade cgu on target hub cluster")
+			})
+
+			By("Deleting rollback cgu created on target hub cluster", func() {
+				err := cnfhelper.DeleteIbuTestCguOnTargetHub(cnfinittools.TargetHubAPIClient, tsparams.RollbackCguName,
+					tsparams.IbuCguNamespace)
+				Expect(err).ToNot(HaveOccurred(), "Failed to delete rollback cgu on target hub cluster")
+			})
+		})
+
+		It("Rollback after failed upgrade", reportxml.ID("69054"), func() {
+			By("Creating, enabling ibu pre-prep CGU and waiting for CGU status to report completed", func() {
+				prePrepCguBuilder := cgu.NewCguBuilder(cnfinittools.TargetHubAPIClient,
+					tsparams.PrePrepCguName, tsparams.IbuCguNamespace, 1).
+					WithCluster(tsparams.TargetSnoClusterName).
+					WithManagedPolicy(tsparams.PrePrepPolicyName).
+					WithCanary(tsparams.TargetSnoClusterName)
+				prePrepCguBuilder.Definition.Spec.Enable = ptr.To(true)
+
+				prePrepCguBuilder, err := prePrepCguBuilder.Create()
+				Expect(err).ToNot(HaveOccurred(), "Failed to create pre-prep CGU.")
+
+				_, err = prePrepCguBuilder.WaitUntilComplete(10 * time.Minute)
+				Expect(err).ToNot(HaveOccurred(), "Pre-prep CGU did not complete in time.")
+			})
+
+			By("Creating, enabling ibu prep CGU and waiting for CGU status to report completed", func() {
+				prepCguBuilder := cgu.NewCguBuilder(cnfinittools.TargetHubAPIClient,
+					tsparams.PrepCguName, tsparams.IbuCguNamespace, 1).
+					WithCluster(tsparams.TargetSnoClusterName).
+					WithManagedPolicy(tsparams.PrepPolicyName).
+					WithCanary(tsparams.TargetSnoClusterName)
+				prepCguBuilder.Definition.Spec.Enable = ptr.To(true)
+
+				prepCguBuilder, err := prepCguBuilder.Create()
+				Expect(err).ToNot(HaveOccurred(), "Failed to create prep CGU.")
+
+				_, err = prepCguBuilder.WaitUntilComplete(25 * time.Minute)
+				Expect(err).ToNot(HaveOccurred(), "Prep CGU did not complete in time.")
+			})
+
+			By("Creating, and enabling ibu upgrade CGU", func() {
+				upgradeCguBuilder := cgu.NewCguBuilder(cnfinittools.TargetHubAPIClient,
+					tsparams.UpgradeCguName, tsparams.IbuCguNamespace, 1).
+					WithCluster(tsparams.TargetSnoClusterName).
+					WithManagedPolicy(tsparams.UpgradePolicyName).
+					WithCanary(tsparams.TargetSnoClusterName)
+				upgradeCguBuilder.Definition.Spec.Enable = ptr.To(true)
+
+				_, err := upgradeCguBuilder.Create()
+				Expect(err).ToNot(HaveOccurred(), "Failed to create upgrade CGU.")
+			})
+
+			By("Performing rollback after a failed upgrade", func() {
+				upgradeStagePolicyStatusCheck, err := ocm.PullPolicy(cnfinittools.TargetHubAPIClient,
+					tsparams.UpgradePolicyName,
+					tsparams.IbuPolicyNamespace)
+				Expect(err).ToNot(HaveOccurred(), "Failed to pull upgrade stage policy")
+
+				err = upgradeStagePolicyStatusCheck.WaitUntilComplianceState(policiesv1.Compliant, 40*time.Minute)
```

Comment:
How do we make the upgrade fail, what's causing it?

Reply:
Good point @mcornea, Assumption here is , upgrade would fail based on code changes in lifecycle-agent operator and test configuration changes, then rollback to previous state root otherwise the SUT/DUT woudn't rollback. If you think, we should simulate a failure to make upgrade failed and perform rollback. i can think through it and add a test step.

Reply:
Just to provide some info for us to consider. In the existing upgrade recovery job, (talm controlled backup and failure recovery using ocp upgrade method), we don't really inject a failure, but rather waiting for the upgrade to proceed to 30%+ and rollback. 

We didn't find a good way to inject failure at that time, and above solution (proposed by Angie) will often/always trigger rollback after etcd was upgraded.

Reply:
To prevent the upgrade from succeeding we can get the machineconfig pool into a degraded state by altering one of the files managed by MCO, e.g. `/etc/mco/proxy.env`. This way we can cover the rollback is automatically triggered in case of failure.

Reply:
Thanks, Yang for sharing more insights. I had a discussion with Marius in this regard. We came up with below proposed methods to trigger auto-rollback upon upgrade failure. 

`
step 1
 By("Creating, enabling ibu pre-prep CGU and waiting for CGU status to report completed",

step2
   By("Creating, enabling ibu prep CGU and waiting for CGU status to report completed"

step3
By("Creating, and enabling ibu upgrade CGU"

step4 
Wait for 15-20 mins (node has rebooted to stateroot B)

step5
By("Verifying auto rollback triggered after a failed upgrade",

_step a_
By("Wait for nodes to become unreachable")

_step b_
By("Wait for nodes to become reachable")

_step c_
By("Wait until all nodes are reporting as Ready")

_step d_
By("Wait for IBU resource to be available")

_step e_
By ("verifying ibu-upgrade-stage-policy status")
      upgradeStagePolicyStatusCheck, err := ocm.PullPolicy(TargetSNOAPIClient,
         tsparams.UpgradePolicyName,
         tsparams.IbuPolicyNamespace)
      Expect(err).ToNot(HaveOccurred(), "Failed to pull upgrade stage policy status")

      err = upgradeStagePolicyStatusCheck.WaitUntilComplianceState(policiesv1.NonCompliant, 3*time.Minute)
      Expect(err).ToNot(HaveOccurred(), "Upgrade-stage-policy failed to report Compliant")

_step f_
 By ("verifying current booted stateroot name") 
_expected result_:
stateroot: rhcos_<seed_image_version>

_step g_
      By("Simulating a fault after node has rebooted into stateroot B", func() {
         faultInjectCmd := "echo a > /etc/mco/proxy.env"  ==> mcp named master will be in "Degraded" state.
         _, err := cluster.ExecCmdWithStdout(TargetSNOAPIClient, faultInjectCmd)
         Expect(err).ToNot(HaveOccurred(), "could not execute command: %s", err)
      })

_step h_
By ("Configuring IBU CR LCA init monitor watchdog timer to 5 mins to trigger auto-rollback")

_step i_
      By ("verifying ibu-upgrade-stage-policy status")", func() {
         upgradeStagePolicyStatusCheck, err := ocm.PullPolicy(TargetSNOAPIClient,
            tsparams.UpgradePolicyName,
            tsparams.IbuPolicyNamespace)
         Expect(err).ToNot(HaveOccurred(), "Failed to pull upgrade stage policy status")

         err = upgradeStagePolicyStatusCheck.WaitUntilComplianceState(policiesv1.NonCompliant,
            tsparams.CustomLcaInitMonitorTimeout=> 5 mins)
         Expect(err).ToNot(HaveOccurred(), "Upgrade-stage-policy expected to report NonCompliant")


step6 (final step)
"Validating target sno cluster version after auto-rollback"
_Expected Result_
Target sno cluster reports correct (previous ocp version) version


`